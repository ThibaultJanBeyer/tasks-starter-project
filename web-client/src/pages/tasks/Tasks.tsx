import React, { useEffect, useState } from 'react'

import * as taskApi from '../../modules/Api'
import { AddTask } from './AddTask'
import { ITask } from '../../types/tasks'
import { TaskItem } from './TaskItem'

interface HandleTaskProps {
  action: keyof typeof taskApi
  data?: ITask | Pick<ITask, '_id'> | any
  setTasks: React.Dispatch<React.SetStateAction<ITask[]>>
}

const handleTasks = async ({ action, data, setTasks }: HandleTaskProps) => {
  try {
    const { tasks } = await taskApi[action](data)
    setTasks(tasks)
  } catch (err) {
    console.log(err)
  }
}

export const Tasks: React.FC = () => {
  const [tasks, setTasks] = useState<ITask[]>([])

  useEffect(() => {
    handleTasks({ action: 'get', setTasks })
  }, [setTasks])

  return (
    <main className="App">
      <h1 style={{ textAlign: 'center' }}>My Tasks</h1>
      <AddTask
        saveTask={(event: React.FormEvent, item: ITask) => {
          event.preventDefault()
          handleTasks({ action: 'add', data: item, setTasks })
        }}
      />
      {tasks.map((task: ITask) => (
        <TaskItem
          key={task._id}
          updateTask={(id) =>
            handleTasks({ action: 'toggle', data: id, setTasks })
          }
          deleteTask={(id) =>
            handleTasks({ action: 'del', data: id, setTasks })
          }
          task={task}
        />
      ))}
    </main>
  )
}
