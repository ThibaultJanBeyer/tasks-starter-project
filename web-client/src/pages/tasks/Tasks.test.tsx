import React from 'react'
import { render, screen } from '@testing-library/react'
import { Tasks } from './Tasks'

test('renders Tasks', () => {
  const { container } = render(<Tasks />)
  const element = screen.getByText(/My Tasks/i)
  expect(element).toBeInTheDocument()
  expect(container).toMatchSnapshot()
})
