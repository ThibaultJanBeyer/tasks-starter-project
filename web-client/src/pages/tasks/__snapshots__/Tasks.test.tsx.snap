// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Tasks 1`] = `
<div>
  <main
    class="App"
  >
    <h1
      style="text-align: center;"
    >
      My Tasks
    </h1>
    <form
      style="border: 1px solid lightgrey; padding: 30px; margin: 10px;"
    >
      <div
        style="display: grid; margin-bottom: 20px;"
      >
        <div
          style="display: grid; grid-template-columns: 100px 1fr; align-items: center;"
        >
          <label
            for="name"
          >
            Name
          </label>
          <input
            id="name"
            type="text"
          />
        </div>
        <div
          style="display: grid; grid-template-columns: 100px 1fr; align-items: center;"
        >
          <label
            for="description"
          >
            Description
          </label>
          <textarea
            id="description"
          />
        </div>
      </div>
      <div
        style="text-align: right;"
      >
        <button
          disabled=""
        >
          Add Task
        </button>
      </div>
    </form>
  </main>
</div>
`;
